### PR TITLE
WT-8417 Restructure assert testing in search near 01 cpp for concurrency

### PR DESCRIPTION
**Pull request by jiechenbo**
Historically search near 01 has suffered through quite alot of concurrency issues when testing the number of expected traversals of a prefix search near. The problem is that testing expects only at max `thread_count` number of search nears to be happening at once. This is a large mistake, because one thread can actually not be scheduled for a long while. 

I have implemented an idea, where we have one read thread that manages the spawning and joining of all threads then checking the statistics. This allows us to be confident that there are no search nears being performed while we are testing the expected number of traversals.

### CODE REVIEW

File: dist/test_data.py [hunk 1]
```diff
@@ -220,7 +220,10 @@ def __ge__(self, other):
                             How long the insertions will occur for.''')]),
     'example_test' : Method(test_config),
     'hs_cleanup' : Method(test_config),
-    'search_near_01' : Method(test_config),
+    'search_near_01' : Method(test_config + [
+                        Config("search_near_threads", 10, r'''
+                            Config that specify the number of search near threads that are
```
Comment (etienneptl) [inline dist/test_data.py:225]: typo `specifies`

Comment (etienneptl) [reply dist/test_data.py:225]: Can we just say something along those lines `Number of threads that execute search near calls`

File: test/cppsuite/configs/search_near_01_default.txt [hunk 2]
```diff
@@ -17,6 +17,6 @@ workload_generator=
     ),
     read_config=
     (
-        thread_count=10
+        thread_count=1
```
Comment (etienneptl) [inline test/cppsuite/configs/search_near_01_default.txt:20]: Can you add a comment explaining why we should not change this number? Add the same comment to the stressing config,

File: test/cppsuite/tests/search_near_01.cxx [hunk 3]
```diff
@@ -170,116 +173,130 @@ class search_near_01 : public test_harness::test {
...
-                tc->transaction.add_op();
-                tc->sleep();
+            /* Cleanup our workers. */
+            for (auto &it : workers) {
+                delete it;
+                it = nullptr;
             }
-            /* Reset our cursor to avoid pinning content. */
-            testutil_check(cursor->reset(cursor.get()));
+
```
Comment (AlexeyAnisimovAU) [inline test/cppsuite/tests/search_near_01.cxx:273]: Do you need to clear `workers` vector after deleting and nullifying every entry in it?

File: test/cppsuite/tests/search_near_01.cxx [hunk 4]
```diff
@@ -170,116 +173,130 @@ class search_near_01 : public test_harness::test {
...
-         * search near to early exit out of it's prefix range when it's trying to search for a
-         * visible key in the tree.
-         */
-        tc->transaction.begin("read_timestamp=" + tc->tsm->decimal_to_hex(10));
+        expected_entries = keys_per_prefix * pow(ALPHABET.size(), PREFIX_KEY_LEN - srchkey_len);
         while (tc->running()) {
+            runtime_monitor::get_stat(
+              tc->stat_cursor, WT_STAT_CONN_CURSOR_NEXT_SKIP_LT_100, &prev_entries_stat);
+            runtime_monitor::get_stat(tc->stat_cursor,
+              WT_STAT_CONN_CURSOR_SEARCH_NEAR_PREFIX_FAST_PATHS, &prev_prefix_stat);
```
Comment (AlexeyAnisimovAU) [inline test/cppsuite/tests/search_near_01.cxx:252]: Minor: we don't have to get stats twice every iteration. Only the first one doesn't have any prev data and that's the only case when we need to get stats at the beginning of the iteration.

**Review by AlexeyAnisimovAU — COMMENTED** [review]
Good work! I have a couple of comments.

File: test/cppsuite/configs/search_near_01_default.txt (hunk 2 above)
Comment (jiechenbo) [reply test/cppsuite/configs/search_near_01_default.txt:24]: Done! Let me know if the comment is good enough.

File: test/cppsuite/tests/search_near_01.cxx (hunk 3 above)
Comment (jiechenbo) [reply test/cppsuite/tests/search_near_01.cxx:276]: Yes we do! Thanks for catching that

File: test/cppsuite/tests/search_near_01.cxx (hunk 4 above)
Comment (jiechenbo) [reply test/cppsuite/tests/search_near_01.cxx:254]: Ahh good point! I actually want to be as confident as we can at this point, in the case that there are search nears performed from WT internally that could potentially increase this stat.

I am fine for both cases here, let me know if we should change it instead.

File: test/cppsuite/tests/search_near_01.cxx [hunk 5]
```diff
@@ -170,116 +173,133 @@ class search_near_01 : public test_harness::test {
...
+        cursor->reconfigure(cursor.get(), "prefix_search=true");
+        /* Generate search prefix key of random length between a -> zzz. */
+        srch_key = random_generator::instance().generate_random_string(
+          srchkey_len, characters_type::ALPHABET);
+        logger::log_msg(LOG_INFO,
+          "Search near thread {" + std::to_string(tc->id) +
+            "} performing prefix search near with key: " + srch_key);
+
+        /*
+         * Read at timestamp 10, so that no keys are visible to this transaction. This allows prefix
```
Comment (etienneptl) [inline test/cppsuite/tests/search_near_01.cxx:193]: I would rephrase the second sentence of that comment:
```
Read at timestamp 10, so that no keys are visible to this transaction and prefix search near early exits when the cursor is positioned on a key that no longer matches the prefix.
```
Also, what if the prefix is `zzz`, will it still be an early exit?

File: test/cppsuite/tests/search_near_01.cxx [hunk 6]
```diff
@@ -170,116 +173,133 @@ class search_near_01 : public test_harness::test {
...
+            logger::log_msg(LOG_INFO,
+              "Read thread skipped entries: " + std::to_string(entries_stat - prev_entries_stat) +
+                " prefix early exit: " +
+                std::to_string(prefix_stat - prev_prefix_stat - z_key_searches));
+            /*
+             * It is possible that WiredTiger increments the entries skipped stat irrelevant to
+             * prefix search near. This is dependent on how many read threads are present in the
+             * test. Account for this by creating a small buffer using thread count. Assert that the
+             * number of expected entries is the upper limit which the prefix search near can
+             * traverse
```
Comment (etienneptl) [inline test/cppsuite/tests/search_near_01.cxx:290]: Full stop missing

File: test/cppsuite/tests/search_near_01.cxx (hunk 4 above)
Comment (AlexeyAnisimovAU) [reply test/cppsuite/tests/search_near_01.cxx:254]: As I said, it's a minor, we can leave it as is. Not a big deal in terms of performance for the test.

**Review by AlexeyAnisimovAU — APPROVED** [review]
Well done! Now the test validation should be more accurate. LGTM!

File: test/cppsuite/tests/search_near_01.cxx [hunk 7]
```diff
@@ -190,9 +190,9 @@ class search_near_01 : public test_harness::test {
             "} performing prefix search near with key: " + srch_key);
 
         /*
-         * Read at timestamp 10, so that no keys are visible to this transaction. This allows prefix
-         * search near to early exit out of it's prefix range when it's trying to search for a
-         * visible key in the tree.
+         * Read at timestamp 10, so that no keys are visible to this transaction. When performing
+         * prefix search near, we expect the search to early exit out of it's prefix range and
```
Comment (etienneptl) [inline test/cppsuite/tests/search_near_01.cxx:194]: Typo `it's`->`its`

**Review by etienneptl — APPROVED** [review]
That looks very good, good work! Please keep executing this specific test for a while before merging it.
